README.md
ジャニーズ当落電話カウンター
============================

## これは何？

ジャニーズファンのための、チケットの確認番号（0570-000-058）に電話をかけ、その電話をかけた回数をカウントするWebアプリです。

さらに、どれくらいで繋がるか、どの時間帯が繋がるか、というデータもちゃっかり収集しちゃいます（※個人情報を収集することはございません）。


## デモ

[デモページはこちら（https://i.y59.jp/Johnny's/callCount/）](https://i.y59.jp/Johnny's/callCount/)


## 注意

- 以下の環境のPHPが必要です。
  - PHP 5.4 +
  - SQLite、SQLite3 関係が動作すること

- スマートフォンで使用されることを想定しています。
  - 正確に言うと、`tel`プロトコルをサポートしている端末からのみ利用できます）。
  - Xperiaなど、一部端末ではレスポンスが遅いと感じることがあります。


## 著作権・ライセンス

© 2016 [yuta*ﾟ](https://y59.jp/), Some Rights Reserved.

ライセンスは [Creative Commons BY-NC 4.0](https://creativecommons.org/licenses/by-nc/4.0/) とします。


CHANGELOG.md
## Version 1.1 (2016-06-18)

Features:

* ユーザーの電話状況を解析出来るようにしました(SQLite3で情報を格納します)。

Changed:

* PHP 5.4+ および SQLite、SQLite3 が使用できる環境が必要になります。


## Version 1.0.2 (2016-06-18)

Fixed:

* Twitter for iPhone アプリ内ブラウザによるバグを修正しました。


## Version 1.0.1 (2016-06-18)

Features:

* アイコンを追加しました。

Fixed:

* ライセンスファイルのインデントがずれていたので、整えました。


## Version 1.0 (2016-06-18)

Features:

* 一通り機能するようになりました。


src/assets/js/script.js
/**
 * script.js
 * ジャニーズ当落電話カウンターのスクリプトファイルです。
 * --------------------
 * @since	1.0
 * @version	1.1
 * @license	Creative Commons BY-NC 4.0
 * 			{@link https://creativecommons.org/licenses/by-nc/4.0/deed.ja}
 * @author	yuta*ﾟ {@link https://y59.jp/}
**/

/****************************
 * 変数定義
*****************************/

// 電話先
var callTo = "0570000058";

// Cookie 処理用
var callArray = [];
var callArray_poped = true;

// Twitter for iPhoneか？未定義の際のフェイルバック
if( typeof twitter_for_iPhone !== "boolean" ) {
	var twitter_for_iPhone = false;
}


/****************************
 * 関数定義
*****************************/

/**
 * クッキーの値を取得
 * @param	cookieName = プロパティ名
 * @return	String
**/
function getCookie(cookieName){
	var dc = document.cookie, st = "", ed = "", rtn = "";
	if(dc.length > 0){
		st = dc.indexOf(cookieName + "=");
		if(st !== -1){
			st = st + parseInt(cookieName.length + 1);
			ed = dc.indexOf(";", st);
			if(ed === -1){
		ed = dc.length;
		}
			// 値をデコードして返す
			rtn = unescape(dc.substring(st, ed));
		}
	} else{
		rtn = "";
	}

	return rtn;
}


/**
 * クッキーを保存
 * @param	cookieName = プロパティ名
 * @param	value = 値
 * @param	expiredays = 有効期限(日数)
 * @return	なし
**/
function setCookie(cookieName, value, expiredays){
	// path の指定
	var path = location.pathname, paths = [];
	// path をフォルダ毎に指定する場合のIE対策
	paths = path.split("/");
	if(paths[paths.length - 1] !== ""){
		paths[paths.length - 1] = "";
		path = paths.join("/");
	};
	// 有効期限の日付
	var extime = new Date().getTime(),
	    cltime = new Date(extime + (86400000 * expiredays)),
	    exdate = cltime.toUTCString();
	// クッキーに保存する文字列を生成
	var s= "";
	s += cookieName + "=" + escape(value);
	s += "; path=" + path;
	if(expiredays){
		s += "; expires=" + exdate + "; ";
	}else{
		s += "; ";
	}
	// クッキーに保存
	document.cookie = s;
}


/**
 * カンマ区切りの挿入。さらに全角数値は半角に変換する
 * @param	destStr = 値
 * @return	カンマ区切りされた値 または undefined
 * @NOTE	min化した時にサイズを減らすため、halfWidthPlus() とセットで作用するようにした
**/
function insertComma(destStr){
	var tmpStr = "";
	if(typeof destStr !== "undefined"){
		destStr = String(destStr);
		destStr = halfWidthPlus(destStr);
		destStr = String(parseFloat(destStr)); // ←念入り
		while (destStr !== (tmpStr = destStr.replace(/^([+-]?\d+)(\d\d\d)/,"$1,$2"))){
			destStr = tmpStr;
		}
		return destStr;
	} else{
		return;
	}
}


/**
 * 全角数字を半角にしたり、不要な文字をトリミングしたりする
 * @param	val = 値
 * @return	トリミングされた値
**/
function halfWidthPlus(val){
	return val.replace(/[Ａ-ｚ０-９]/g, function(s){
		return String.fromCharCode(s.charCodeAt(0) - 0xFEE0);
	}).replace(/(,|，|､|、|\ |\　)/g,"").replace(/．/g,".").replace(/\‐\-\‑\—/g,"-");
}


/**
 * カウンターを設定する
 * @param	setValue = 値
 * @param	relative = 相対的に値を増減するか
 * @return	Number(設定後の値) または undefined
**/
function counterSet(setValue, relative){
	// Init
	var current = parseInt($("#callValue").attr("data-value")), setCount;
	// relative が true の場合、相対的に値を増減する
	if(relative){
		setCount = current + setValue;
	} else{
		setCount = setValue;
	}
	// ただし、setCount が 0 以上の場合のみ処理をする
	if(setCount >= 0){
		// 表示値・アトリビュートを変える
		counterAffect(setCount);
		// Cookie に値を格納する
		setCookie("callCount_value", setCount, 731);
		return setCount;
	} else{
		return;
	}
}


/**
 * カウンターの表示を変える
 * @NOTE	主に counterSet() とセットで使用する
 * @param	val = 値
 * @return	なし
**/
function counterAffect(val){
	$("#callValue").attr("data-value", String(val)).html(insertComma(val));

	// 繋がった! ボタンの表示・非表示を切り替える
	var outWentButton = $(".outWent");
	if( parseInt(val) >= 1 ){
		outWentButton.show();
	} else{
		outWentButton.hide();
	}
}


/**
 * callArray 配列に、現在の時間(Unixミリ秒)を入れる
 * @param	なし
 * @return	なし
 * @since	Version 1.1
**/
function callArray_push(){
	callArray.push(String(new Date().getTime()));
	callArray_poped = false;
	callArray_save();
}

/**
 * callArray 配列から、後方1つを取り除く
 * @param	なし
 * @return	なし
 * @since	Version 1.1
**/
function callArray_pop(){
	if(! callArray_poped){
		callArray.pop();
		callArray_poped = true;
		callArray_save();
	};
}

/**
 * callArray 配列を Cookie に入れる
 * @param	なし
 * @return	なし
 * @since	Version 1.1
**/
function callArray_save(){
	var callArray_string = callArray.join(",");
	setCookie("callCount_array", callArray_string, 731);
}


/****************************
 * 即時処理
*****************************/

// Cookie から ajax 送信用の途中データがあったら、配列に格納する
var _ic = getCookie("callCount_array");
if(_ic !== "" && typeof _ic !== "undefined"){
	callArray = _ic.split(",");
}


/****************************
 * 即時関数
*****************************/

$(function($doc){

	// タッチイベントを扱えるかどうかによって、イベントハンドラを変える
	var supportTouch = "ontouchend" in document;
	var EVENT_TOUCHSTART = supportTouch ? "touchstart": "mousedown";

	// Cookie から電話カウンターを読み込む。読み込めたら値を表示する
	// @NOTE: PHP 側と重複しているが、JavaScript 側でも処理
	var initValue = getCookie("callCount_value");
	if(initValue !== "" && typeof initValue !== "undefined"){
		counterAffect(initValue);
	}

	// 電話をかける & カウンターを1増やす
	$($doc).on(EVENT_TOUCHSTART, ".callButton", function(){
		location.href = "tel:" + callTo;
		counterSet(1, true);
		callArray_push();
	});

	// 電話カウンターを調整する
	$($doc).on(EVENT_TOUCHSTART, ".adjust", function(){
		var attrVal = $(this).attr("data-count");
		switch (attrVal){
			case "+1":
				// カウンターを1増やす
				counterSet(1, true);
				break;
			case "-1":
				// カウンターを1減らす
				counterSet(-1, true);
				callArray_pop();
				break;
			case "custom":
				// カウンターをカスタム調整する
				var newCount = window.prompt("電話をかけた回数を半角で入力してください", parseInt($("#callValue").attr("data-value")));
				if(typeof newCount !== "undefined" && newCount !== "" && newCount !== null){
					newCount = halfWidthPlus(newCount);
					var newCountInteger = parseInt(newCount);
					// 正しい値かを調査する
					if(newCount === String(newCountInteger)){
						// Valid Value
						if(newCount >= 0){
							counterSet(newCount, false);
							setCookie("callCount_array", "", -1);
						} else{
							alert("注：0 以上の数字を入力してください。");
						}
					} else{
						// Invalid Value
						alert("注：正しい数字を入力してください。");
					}
				}
				break;
			case "reset":
				// カウンターをリセットする
				if(window.confirm("電話カウンターをリセットしてもよろしいですか?")){
					counterSet(0, false);
					setCookie("callCount_array", "", -1);
				}
				break;
			default:
				// Do Nothing
				break;
		}
	});

	// 繋がった処理
	$($doc).on(EVENT_TOUCHSTART, ".outWent", function(e){

		// 繋がった処理をするかどうかのブーメランを予め定義する
		var outWentProcess = false;
		if(twitter_for_iPhone){
			outWentProcess = true;
		} else {
			if(window.confirm("当落確認が出来ましたか?")){
				outWentProcess = true;
			}
		}

		if(outWentProcess){
			// .notConnected をフェードアウト → .connected をフェードイン
			var elem1 = $(".notConnected"), elem2 = $(".connected");
			elem1.addClass("fadeout");
			setTimeout(function(){
				elem1.hide();
				elem2.addClass("fadein");
			}, 500);
			setTimeout(function(){
				elem2.show();
			}, 1000);

			// Ajaxで情報送信
			$.ajax({
				type: "POST",
				url: "ajaxdata.php",
				data: {
					"counter": $("#callValue").attr("data-value"),
					"date_array": callArray,
					"date_send": String(new Date().getTime()),
				},
				timeout: 10,
				dataType: "json",
				async: false, // true にすると情報が飛ばなくなる
			});

			// Cookie内のカウンターリセット
			setCookie("callCount_value", "0", -1);
			setCookie("callCount_array", "", -1);
			callArray = [];
		}

	});

	// シェアボタンを押された処理
	$($doc).on(EVENT_TOUCHSTART, ".share", function(){
		var shareType = $(this).attr("data-shareto"),
		    myCount = $("#callValue").attr("data-value"),
		    thisURL = location.href,
		    shareText = encodeURI("ジャニーズ当落電話カウンター! 私は" + insertComma(myCount) + "回で繋がりました٩(๑^o^๑)۶"),
			shareHashTag = encodeURI("当落電話カウンター"),
			shareURL = "";
		switch (shareType){
			case "twitter":
				shareURL = "https://twitter.com/share?via=y59JP&text=" + shareText + "&url=" + thisURL + "&hashtags=" + shareHashTag + "&lang=ja";
				break;
			case "facebook":
				shareURL = "https://www.facebook.com/sharer.php?u=" + thisURL;
				break;
			case "line":
				shareURL = "line://msg/text/" + shareText + "%0D%0A" + encodeURI(thisURL);
				break;
			default:
				break;
		};
		if(shareURL !== "") {
			if(twitter_for_iPhone){
				location.href = shareURL;
			} else {
				window.open(shareURL, "_blank");
			};
		}
	});

}($(document)));


src/ajaxdata.php
<?php
/**
 * ajaxdata.php
 * ジャニーズ当落カウンターの使用状況を送信します。
 * @NOTE	必要な SQLite3 テーブルは既に出来ている体になっています。
 * @NOTE	Require PHP 5.4+, SQLite, SQLite3
 * --------------------
 * @param 	counter: カウンター
 * @param	date_array: 日付情報(Unixミリ秒・配列)
 * @param	date_send: このAjaxの送信日付(Unixミリ秒)
 *
 * @return	array( "finished" => Bool, [ "errorReason" => エラーになった理由 ] )
 * --------------------
 * @since	1.1
 * @version	1.1
 * @license	Creative Commons BY-NC 4.0
 * 			{@link https://creativecommons.org/licenses/by-nc/4.0/deed.ja}
 * @author	yuta*ﾟ {@link https://y59.jp/}
**/

// 実稼働環境：Fatal Error以外のエラーを非表示にする
error_reporting( E_ERROR );

// ヘッダー
header( "Content-Type: application/json; charset=UTF-8" );

// 戻り値
$return = array(
	"finished" => false,
);

// 取得した情報を配列に入れる
$counter    = $_REQUEST["counter"];
$date_array = $_REQUEST["date_array"];
$date_send  = $_REQUEST["date_send"];

// 必要な情報がちゃんとあるか調べる
if( (string)$counter !== "" && @is_array( $date_array ) && @count( $date_array ) >= 1 && (string)$date_send !== "" ) {

	// DBに格納するように日付データを作る
	$sql_date_array = array();
	$lastKey = 0;
	foreach( $date_array as $key => $value ) {
		// ミリ秒ではなく、秒を出す
		$date_seconds = floor( (int)$value / 1000 );
		// 配列に格納する
		$sql_date_array[] = array(
			"date_str" => date( "Y-m-d H:i:s", $date_seconds ),
			"date_int" => $value
		);
		// 最後の時間を取得するため、最後のキー番号を変数に入れる
		$lastKey = $key;
	};

	// データの送信日もDBで処理しやすい形式にする
	$sql_date_send_int = floor( (int)$date_send / 1000 );
	$sql_date_send     = date( "Y-m-d H:i:s", $sql_date_send_int );

	// SQLite 接続
	try{
		$SQLite = new SQLite3( "assets/data.sqlite3" );
	} catch( Exception $e ) {
		$return["errorReason"] = "DBの接続に失敗しました({$e})";
		die( json_encode( $return, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE ) );
	}

	// ユーザー情報を入れる
	$user_SQL =
		"INSERT INTO `user` (
			`count`,
			`userAgent`,
			`date_1st`,
			`date_1st_int`,
			`date_final`,
			`date_final_int`,
			`date_datasend`,
			`date_datasend_int`
		 ) VALUES (
			 '{$counter}',
			 '{$_SERVER["HTTP_USER_AGENT"]}',
			 '{$sql_date_array[0]["date_str"]}',
			 '{$sql_date_array[0]["date_int"]}',
			 '{$sql_date_array[$lastKey]["date_str"]}',
			 '{$sql_date_array[$lastKey]["date_int"]}',
			 '{$sql_date_send}',
			 '{$date_send}'
		 )";
	$user_result = $SQLite -> query( $user_SQL );
	$user_lastID = $SQLite -> lastInsertRowId();

	// 電話情報を入れる
	$call_SQL = "INSERT INTO `call` (`userID`, `date`, `date_int`, `connect`) VALUES \n";
	foreach( $sql_date_array as $key => $value ) {
		$finalLoop = $value === end( $sql_date_array ) ? true : false;
		$connect = $finalLoop ? "1" : "0";
		$call_SQL .= "('{$user_lastID}', '{$value["date_str"]}', '{$value["date_int"]}', '{$connect}')";
		if(! $finalLoop ) {
			$call_SQL .= ", \n";
		}
	};
	$call_result = $SQLite -> query( $call_SQL );

	// $call_result が false でなければ、正しく処理したものとして返り値を true にする
	if( $call_result !== false ) {
		$return["finished"] = true;
	} else {
		$return["errorReason"] = "DBに値を格納できなかった可能性があります";
	}

	// SQLite を閉じる
	$SQLite -> close();

} else {
	// 必要な情報が無かった場合
	$return["errorReason"] = "必要な情報がありません";
}

// クライアントサイドに返却する
echo json_encode( $return, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE );

?>
